libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1350,16 +1359,62 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 		m_context << Instruction::SUB;
 		break;
 	case Token::Mul:
-		if (c_isFractional)
+		if (c_isFractional) //fixed point handling
 		{
-			//multiply, then shift right...this is your fraction.
-			m_context << Instruction::MUL << c_shift << Instruction::SWAP1 << Instruction::DIV;
-			//now redo the process...
-			m_context << Instruction::DUP2 << Instruction::DUP4;
-			//but this time, get the integer portion.
-			m_context << c_shift << Instruction::SWAP1 << Instruction::DIV << Instruction::MUL;
-			//add
-			m_context << Instruction::ADD;
+			if (c_numBits - c_fractionalBits == 0)
+			{
+				//take two numbers, cut them in half, multiply them, simple as that.
+				m_context << c_halfShift << Instruction::DUP1 << Instruction::SWAP2 << (c_isSigned ? Instruction::SDIV : Instruction::DIV); 
+				m_context << Instruction::SWAP2 << (c_isSigned ? Instruction::SDIV : Instruction::DIV) << Instruction::MUL;
+			}
+			else if (c_numBits > 128)
+			{
+				//duplicate converted types for reusage
+				m_context << Instruction::DUP1 << Instruction::DUP3;
+				
+				//time to get schwifty in here...
+				//split both numbers into their representative fractional and decimal parts
+				//this takes the form A.B * C.D
+				//then we utilize this formula:
+				//D*B/shift + C*A*shift + D*A + C*B 
+				
+				//D*B/shift				
+				m_context << c_fractionShift << Instruction::SWAP1 << (c_isSigned ? Instruction::SMOD : Instruction::MOD);
+				m_context << Instruction::SWAP1 << c_fractionShift <<  Instruction::SWAP1 << (c_isSigned ? Instruction::SMOD : Instruction::MOD);
+				if (c_fractionalBits > c_intBits)
+				{
+					//we need this here because overwise we will have an overflow with our fractional bits
+					m_context << c_halfShift << Instruction::SWAP1 << (c_isSigned ? Instruction::SDIV : Instruction::DIV);
+					m_context << Instruction::SWAP1 << c_halfShift << Instruction::SWAP1 << (c_isSigned ? Instruction::SDIV : Instruction::DIV) << Instruction::MUL;
+				}
+				else
+					m_context << Instruction::MUL << c_fractionShift << Instruction::SWAP1 << (c_isSigned ? Instruction::SDIV : Instruction::DIV);
+				//C*A*shift*/
+				m_context << Instruction::DUP2 << Instruction::DUP4;
```

Comment:
what do you mean by using inline assembly here, is that not what we are doing? Or are you saying users should be using inline assembly here?